pyteal/compiler/compiler_test.py

```diff
@@ -2326,25 +2324,8 @@ def log_creation(*, output: pt.abi.String) -> pt.Expr:
         )
         assert meth.method_signature() == "log_creation()string"
 
-        @pt.ABIReturnSubroutine
-        def approve_if_odd(condition_encoding: pt.abi.Uint32) -> pt.Expr:
```

Comment:
This deleted router method is a great example to think through. 
In the short term, I support the current strict approach. But in the medium term, it's worth seeing if we can relax the restrictions a bit.
As we discussed, there could still be a use case when someone actually wants to _fail_ during a clear state cleanup action. So the `approve_if_odd` can be a stand-in for that. How would we want to enable such a call? Some ideas:
* Do we allow any arguments at all?
* If so, what kind of checks for the arguments should we generate? In particular, what happens if the 0'th argument - for the method selector - is incorrect?
* Should we allow an option that instead of issuing `err` issues `int 1; return` immediately instead?

These questions get kind of dicey, but I can imagine modifying `@router.method` (or maybe even introducing `@router.clear_method`) with some parameters that control this. EG:
* `for_clear: bool` (unneeded in the case of `@router.clear_method`) indicates whether this is intended for the clear program, and then enables the following parameters:
* `no_erroring = True` - in the case that the method selector (or any other parameter) is wrong, will issue an `int 1; return` instead of `err`. But this means that all the other `@router.clear_method` would need to have the same `no_erroring` value. So maybe the _routing err_ behavior belongs as a _router level option_.

Reply:
I think my take on the discussion is that, we might want to let router user to *explicitly* specify what kind of logic they want to use in clear state program (CSP), so my thoughts and opinions are as follows:

> Do we allow any arguments at all?

There is no way we are disallowing args, one can always use `Txn.application_args` in their composed `Expr`. We will just no longer generate code to handle the IO process in CSP.

> If so, what kind of checks for the arguments should we generate? In particular, what happens if the 0'th argument - for the method selector - is incorrect?

If we go down the path of letting users to specify whatever logic in CSP, then I don't see too good a way to generate checks for arguments.

For now, I haven't thought through how much flexibility and check can be done on our side, but rather, probably we want to hand the full control over CSP to router users?